### PR TITLE
adding teachers to school interaction

### PR DESCRIPTION
**Pull request by terasaki1**
Adults were not initially assigned to school, so the interactions between adults and children were not captured in [InteractionModSchool.H.](https://github.com/AMReX-Codes/ExaEpi/blob/81f6b5077cb8f463fa9e5586c67ba7b5bce20b85/src/InteractionModSchool.H#L18)

### CODE REVIEW

File: src/AgentContainer.H [hunk 1]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (debog) [inline src/AgentContainer.H]: remove extra empty line

File: src/AgentContainer.cpp [hunk 2]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (debog) [inline src/AgentContainer.cpp]: remove extra empty line

File: src/AgentContainer.cpp [hunk 3]
```diff
@@ -518,6 +535,10 @@ void AgentContainer::initAgentsCensus (iMultiFab& num_residents,    /*!< Number
         auto dx = ParticleGeom(0).CellSizeArray();
         auto my_proc = ParallelDescriptor::MyProc();
 
+        auto student_counts_arr = student_counts[mfi].array();
+        auto teacher_counts_arr = teacher_counts[mfi].array();
+
+
```
Comment (debog) [inline src/AgentContainer.cpp:543]: remove extra empty line

File: src/AgentContainer.cpp [hunk 4]
```diff
@@ -636,6 +657,7 @@ void AgentContainer::initAgentsCensus (iMultiFab& num_residents,    /*!< Number
                 work_nborhood_ptr[ip] = 5*nborhood;
                 workgroup_ptr[ip] = 0;
 
+
```
Comment (debog) [inline src/AgentContainer.cpp]: remove extra empty line

File: src/DiseaseParm.H [hunk 5]
```diff
@@ -43,7 +43,12 @@ struct DiseaseParm
     /*! Transmission probabilities of children for each age group (<5, 5-17, 18-29, 30-64, 65+) for neighborhood cluster */
     amrex::Real xmit_nc_child[5] = {amrex::Real(.15*pHCC0), amrex::Real(.15*pHCC1), amrex::Real(.08*pHC), amrex::Real(.08*pHC), amrex::Real(.08*pHC)};
     /*! Work transmission probabilities for each age group (<5, 5-17, 18-29, 30-64, 65+) */
+    /* adding hypothetical transmission for now for work interaction, if a_j is older, will transmission be higher?*/
     amrex::Real xmit_work = amrex::Real(0.115*pWO);
+    amrex::Real xmit_work_diff_age = amrex::Real(0.090*pWO);
+    amrex::Real xmit_work_diff_group = amrex::Real(0.075*pWO);
+    amrex::Real xmit_work_diff_group_diff_age = amrex::Real(0.060*pWO);
+
```
Comment (debog) [inline src/DiseaseParm.H]: Do these lines belong to this PR? Or is this from the work you are doing with improving workplace interactions?

File: src/DiseaseParm.cpp [hunk 6]
```diff
@@ -33,6 +33,9 @@ void DiseaseParm::Initialize ()
     xmit_nc_child[2] = xmit_nc_child[3] = xmit_nc_child[4] = .08_rt*pHC;
 
     xmit_work = 0.115_rt*pWO;
+    xmit_work_diff_age = 0.090_rt*pWO;
+    xmit_work_diff_group = 0.075_rt*pWO;
+    xmit_work_diff_group_diff_age = 0.060_rt*pWO;  // later model, when a_j is older than a_i
```
Comment (debog) [inline src/DiseaseParm.cpp]: Do these lines belong to this PR? Or is this from the work you are doing with improving workplace interactions?

File: src/DiseaseParm.cpp [hunk 7]
```diff
@@ -65,6 +68,9 @@ void DiseaseParm::Initialize ()
 
     // Multiply contact rates by transmission probability given contact
     xmit_work *= p_trans[0];
+    xmit_work_diff_age *= p_trans[0];
+    xmit_work_diff_group *= p_trans[0];
+    xmit_work_diff_group_diff_age *= p_trans[0];
```
Comment (debog) [inline src/DiseaseParm.cpp]: Do these lines belong to this PR? Or is this from the work you are doing with improving workplace interactions?

File: src/AgentContainer.H [hunk 8]
```diff
@@ -293,8 +299,37 @@ public:
...
 
+    amrex::iMultiFab student_counts;
+    amrex::iMultiFab teacher_counts;
+    amrex::Gpu::DeviceVector<int> Unit_total_teacher_counts;
+    amrex::Gpu::DeviceVector<int> Unit_elem3_teacher_counts;
+    amrex::Gpu::DeviceVector<int> Unit_elem4_teacher_counts;
+    amrex::Gpu::DeviceVector<int> Unit_midl_teacher_counts;
+    amrex::Gpu::DeviceVector<int> Unit_high_teacher_counts;
+    amrex::Gpu::DeviceVector<int> Unit_daycr_teacher_counts;
+
```
Comment (debog) [inline src/AgentContainer.H]: Is it possible to have a 2D vector instead of separate variable for each teacher category, where one index will indicate category (total, elem, midl, high, etc) and the other will indicate unit?

File: src/AgentContainer.cpp [hunk 9]
```diff
@@ -644,6 +666,49 @@ void AgentContainer::initAgentsCensus (iMultiFab& num_residents,    /*!< Number
...
+        amrex::ParallelFor(bx,
+            [=] AMREX_GPU_DEVICE (int i, int j, int k) noexcept
+        {
+            if (k == 0){ // avoid double compute
+                teacher_counts_arr(i, j, 0) = (student_counts_arr(i, j, 0) + 19)  / 20;
+                teacher_counts_arr(i, j, 1) = (student_counts_arr(i, j, 1) + 19)  / 20;
+                teacher_counts_arr(i, j, 2) = (student_counts_arr(i, j, 2) + 19)  / 20;
+                teacher_counts_arr(i, j, 3) = (student_counts_arr(i, j, 3) + 19)  / 20;
+                teacher_counts_arr(i, j, 4) = (student_counts_arr(i, j, 4) + 19)  / 20;
```
Comment (debog) [inline src/AgentContainer.cpp]: Instead of hard-coding `20`, we can instead have an int vector of ratios. The default can be `[20 20 20 20 20]`, but can also be read in from input file?

File: src/Initialization.cpp [hunk 10]
```diff
@@ -193,18 +223,76 @@ namespace Initialization
...
+                                workgroup_ptr[ip] = 1 ;
+                                amrex::Gpu::Atomic::AddNoRet(d_high_teachers, 1);
+                            } else if (choice < total_available) {
+                                school_ptr[ip] = 5;  // day care
+                                workgroup_ptr[ip] = 5 ;
+                                amrex::Gpu::Atomic::AddNoRet(d_daycr_teachers, 1);
+                            }
+                        }
+
+
```
Comment (debog) [inline src/Initialization.cpp]: remove extra empty line

File: src/InteractionModWork.H
Comment (debog) [inline src/InteractionModWork.H:1]: These commits shouldn't belong to this PR, right?

**Comment by debog** [comment]
@terasaki1 Is there a good way to do a sanity check for correctness? Once you add teachers, the infection trends will change. But, say you specify the ratio of students to teachers as something very large instead of `20`, then the code should compute 0 teachers, and you should get _exactly_ the same solution as the `development` branch. Can you try this?

@atmyers @tannguyen153 Do you think this is a "necessary and sufficient" test?

File: src/AgentContainer.cpp [hunk 11]
```diff
@@ -644,6 +666,49 @@ void AgentContainer::initAgentsCensus (iMultiFab& num_residents,    /*!< Number
                     school_ptr[ip] = -1;
                 }
 
+                // Increment the appropriate student counter based on the school assignment
+                if (school_ptr[ip] == 3) {
+                    amrex::Gpu::Atomic::AddNoRet(&student_counts_arr(i, j, 3), 1);  // Elementary school 3
```
Comment (debog) [inline src/AgentContainer.cpp]: shouldn't it be `students_counts_arr(i,j,k,3)` instead of `students_counts_arr(i,j,3)`? The `3` here is the component index for the `MultiFab`, yes? Not the 3rd dimension?
(same for the other places where you access `student_counts` and `teacher_counts`)

File: src/AgentContainer.cpp [hunk 12]
```diff
@@ -644,6 +666,49 @@ void AgentContainer::initAgentsCensus (iMultiFab& num_residents,    /*!< Number
...
+                } else if (school_ptr[ip] == 4) {
+                    amrex::Gpu::Atomic::AddNoRet(&student_counts_arr(i, j, 4), 1);  // Elementary school 4
+                } else if (school_ptr[ip] == 2) {
+                    amrex::Gpu::Atomic::AddNoRet(&student_counts_arr(i, j, 2), 1);  // Middle school
+                } else if (school_ptr[ip] == 1) {
+                    amrex::Gpu::Atomic::AddNoRet(&student_counts_arr(i, j, 1), 1);  // High school
+                } else if (school_ptr[ip] == 5) {
+                    amrex::Gpu::Atomic::AddNoRet(&student_counts_arr(i, j, 0), 1);  // Playgroups + Day care
+                }
```
Comment (debog) [inline src/AgentContainer.cpp]: Instead of hardcoding component numbers `0, 1, ..., 4` at all the places (and consequently introducing the prospect of making a mistake), can you create an `enum` type (for example, like [`Status`](https://github.com/AMReX-Codes/ExaEpi/blob/81f6b5077cb8f463fa9e5586c67ba7b5bce20b85/src/AgentDefinitions.H#L71)?

```
struct SchoolType
{
    enum {
        day_care = 0,
        high,
        middle,
        elem_3,
        elem_4,
        total
    };
};
```

Then you can use `student/teacher_counts_arr(i,j,k,SchoolType::high)`, etc.

File: src/DiseaseParm.H (hunk 5 above)
Comment (terasaki1) [reply src/DiseaseParm.H]: true. Ill remove them

File: src/DiseaseParm.cpp (hunk 6 above)
Comment (terasaki1) [reply src/DiseaseParm.cpp]: true. I will remove them

File: src/AgentContainer.cpp (hunk 11 above)
Comment (terasaki1) [reply src/AgentContainer.cpp]: correct. it is the third element. Since the grid is in 2D. the k_th element is ignored. However, in the code some multifab are defined as fab(i,j,k,n) and some as fab(i,j,n).

I will change it to (i,j,k,n) for generalization

File: src/AgentContainer.cpp (hunk 9 above)
Comment (terasaki1) [reply src/AgentContainer.cpp]: ratio can be read from input file now as agent.teacher_student_ratios

File: src/AgentContainer.H (hunk 8 above)
Comment (terasaki1) [reply src/AgentContainer.H]: amrex::Gpu::DeviceVector<int>amrex::Gpu::DeviceVector<int> does not support 2D. Should I leave it or use flatten 1D vector?

File: src/AgentContainer.H [hunk 13]
```diff
@@ -74,20 +74,27 @@ public:
...
+        teacher_counts.setVal(0);
+
         add_attributes();
 
         {
             amrex::ParmParse pp("agent");
             pp.query("symptomatic_withdraw", m_symptomatic_withdraw);
             pp.query("shelter_compliance", m_shelter_compliance);
             pp.query("symptomatic_withdraw_compliance", m_symptomatic_withdraw_compliance);
+            pp.queryarr("teacher_student_ratios", teacher_student_ratios);
```
Comment (debog) [inline src/AgentContainer.H]: Given how you are using this variable, change it to `student_teacher_ratios` since it represents student to teacher ratio, not the teacher to student ratio.

File: src/AgentDefinitions.H [hunk 14]
```diff
@@ -67,6 +67,20 @@ struct IntIdxDisease
...
 };
 
+/*! \brief School Type  */
+struct SchoolType
+{
+    enum {
+        total = 0,   /*!< All School type */
+        high = 1,    /*!< High School */
+        middle,      /*!< Middle School */
+        elem_3,      /*!< Elementary School Neighbordhoo 1 */
```
Comment (atmyers) [inline src/AgentDefinitions.H]:
```suggestion
        elem_3,      /*!< Elementary School Neighborhood 1 */
```

File: src/AgentDefinitions.H [hunk 15]
```diff
@@ -67,6 +67,20 @@ struct IntIdxDisease
...
+        elem_4,      /*!< Elementary School Neighbordhoo 1 */
```
Comment (atmyers) [inline src/AgentDefinitions.H]:
```suggestion
        elem_4,      /*!< Elementary School Neighborhood 1 */
```

File: src/AgentDefinitions.H [hunk 16]
```diff
@@ -67,6 +67,20 @@ struct IntIdxDisease
     };
 };
 
+/*! \brief School Type  */
+struct SchoolType
+{
+    enum {
+        total = 0,   /*!< All School type */
+        high = 1,    /*!< High School */
```
Comment (atmyers) [inline src/AgentDefinitions.H]:
```suggestion
        high,    /*!< High School */
```

File: src/DiseaseParm.H [hunk 17]
```diff
@@ -44,7 +44,7 @@ struct DiseaseParm
     amrex::Real xmit_nc_child[5] = {amrex::Real(.15*pHCC0), amrex::Real(.15*pHCC1), amrex::Real(.08*pHC), amrex::Real(.08*pHC), amrex::Real(.08*pHC)};
     /*! Work transmission probabilities for each age group (<5, 5-17, 18-29, 30-64, 65+) */
     amrex::Real xmit_work = amrex::Real(0.115*pWO);
-    /*! School transmission probabilities for each age group (<5, 5-17, 18-29, 30-64, 65+) */
+    /*! School transmission probabilities for each school tyoe (High, Middle, Elem3/4 , Daycare/Playgroups */
```
Comment (atmyers) [inline src/DiseaseParm.H]:
```suggestion
    /*! School transmission probabilities for each school type (High, Middle, Elem3/4 , Daycare/Playgroups */
```

File: src/AgentContainer.H [hunk 18]
```diff
@@ -74,20 +74,25 @@ public:
...
+        student_counts.setVal(0);  // Initialize the MultiFab to zero
+
         add_attributes();
 
         {
             amrex::ParmParse pp("agent");
             pp.query("symptomatic_withdraw", m_symptomatic_withdraw);
             pp.query("shelter_compliance", m_shelter_compliance);
             pp.query("symptomatic_withdraw_compliance", m_symptomatic_withdraw_compliance);
+            pp.queryarr("student_teacher_ratios", student_teacher_ratios);
```
Comment (debog) [inline src/AgentContainer.H:94]: @terasaki1 Can you add documentation for this new input parameter [here](https://github.com/AMReX-Codes/ExaEpi/blob/41e4f3b953ac5a40d38803ecde4a09239085e832/docs/source/usage/how_to_run.rst?plain=1#L93)? Just follow the example for the other input parameters; this will then appear [here](https://exaepi.readthedocs.io/en/latest/usage/how_to_run.html#inputs-parameters).

**Review by debog — APPROVED** [review]

**Review by atmyers — APPROVED** [review]

**Comment by atmyers** [comment]
Thank you for this!
